### PR TITLE
Release v1.6.0 #570

### PR DESCRIPTION
**Pull request by MichaelStritt**
Check out the release issue #570.

### CODE REVIEW

**Comment by HenkMutsaerts** [comment]
Warning: Field EchoTime in xASL structure has a value
0.01056, which is outside of the recommended range <0.5,500>
ms. 
> In xASL_bids_parms2BIDS (line 107)
  In xASL_bids_Dicom2JSON (line 619)
  In xASL_imp_DCM2NII_Subject_StoreJSON (line 38)
  In xASL_imp_DCM2NII_Subject (line 206)
  In xASL_imp_DCM2NII (line 252)
  In xASL_module_Import (line 203)
  In xASL_test_BIDSConversion (line 66)
  In xASL_test_BIDSFlavorsFull (line 119) 
ASL context: ,
Sorted NIfTIs based on InstanceNumbers...
Warning: concatenating multiple NIfTIs & jsons as output from dcm2niiX
Corrected dcm2niiX output for following files:
/Users/henk/ExploreASL/ASL/TestBIDS/TmpConversion/GE_PCASL_3Dspiral_Product_pharma/analysis/11/ASL_1/ASL4D.nii

dataset_description.json is missing the following RECOMMENDED fields: 
HEDVersion, DatasetType, License
Reference to non-existent field 'ASLContext'.

Error in xASL_imp_NII2BIDS_Subject_DefineM0Type (line 64)
                elseif
                ~isempty(strfind(jsonLocal.ASLContext,bidsPar.strM0scan))
                
Error in xASL_imp_NII2BIDS_SubjectSessionRun (line 345)
        xASL_imp_NII2BIDS_Subject_DefineM0Type(studyPar,
        bidsPar, jsonLocal, inSessionPath, subjectLabel,
        sessionLabel);

Error in xASL_imp_NII2BIDS_SubjectSession (line 69)
            imPar, bidsPar, studyPar, subjectLabel,
            sessionLabel, listSubjects, fSes, inSessionPath,
            outSessionPath, nSes, iSubject, kk, mm);

Error in xASL_imp_NII2BIDS_Subject (line 90)
            xASL_imp_NII2BIDS_SubjectSession(imPar, bidsPar,
            studyPar, iSubject, fSes, listSubjects,
            subjectLabel, kk);

Error in xASL_imp_NII2BIDS (line 67)
        xASL_imp_NII2BIDS_Subject(imPar,bidsPar,studyPar,listSubjects,iSubject);
        
Error in xASL_module_Import (line 209)
        xASL_imp_NII2BIDS(imPar, studyPath, studyParPath);

Error in xASL_test_BIDSConversion (line 136)
        xASL_module_Import(fullfile(baseDirImport,
        flavorList{iFlavor}), [],[], [0 1 0], false, true,
        false, false, x);

Error in xASL_test_BIDSFlavorsFull (line 119)
	xASL_test_BIDSConversion(conversionPath,
        referencePath, 1, 0);

**Comment by MichaelStritt** [comment]
@HenkMutsaerts which dataset/os/etc. ?

**Comment by jan-petr** [comment]
> @HenkMutsaerts which dataset/os/etc. ?

Stupid MacOS is my guess :D

**Comment by jan-petr** [comment]
> > @HenkMutsaerts which dataset/os/etc. ?
> 
> Stupid MacOS is my guess :D

Because it runs dicominfo instead of DCMTK (that doesn't work because of a compilation issue). So we need to add all the new extra DICOM fields to dicominfo reading that we have added recently to the DCMTK MEX file... Lemme do that...

**Comment by jan-petr** [comment]
Hm - so there should be any problem on MacOS either... Are all these errors all from GE_PCASL_3Dspiral_Product_pharma?

I only get
`dataset_description.json is missing the following RECOMMENDED fields:
HEDVersion, DatasetType, License`

Which is an OK BIDS warning. Have you updated the Flavors?

**Comment by HenkMutsaerts** [comment]
> Hm - so there should be any problem on MacOS either... Are all these errors all from GE_PCASL_3Dspiral_Product_pharma?
> 
> I only get
> `dataset_description.json is missing the following RECOMMENDED fields: HEDVersion, DatasetType, License`
> 
> Which is an OK BIDS warning. Have you updated the Flavors?

Yes from GE pharma indeed. Flavors are updated to last main. So there is no dicominfo  or dcmtk related issue? What was the reason again that we couldn't compile dcmtk for mac?

**Comment by HenkMutsaerts** [comment]
@jan-petr it goes wrong in xASL_imp_NII2BIDS_Subject_DefineM0Type as it tries to read `jsonLocal.ASLContext`, but this field doesn't exist. Should this field exist already at this point? Or should it be a separate file? If I look at the folder, only a NIfTI exists with two images (ASL & M0).

**Comment by MichaelStritt** [comment]
> 
> 
> @jan-petr it goes wrong in xASL_imp_NII2BIDS_Subject_DefineM0Type as it tries to read `jsonLocal.ASLContext`, but this field doesn't exist. Should this field exist already at this point? Or should it be a separate file? If I look at the folder, only a NIfTI exists with two images (ASL & M0).

So if I pause directly at the beginning of `xASL_imp_NII2BIDS_Subject_DefineM0Type` I get:

image

In the rawdata folder, there should only be the NIFTI. Within the analysis folder, there should still be the JSON.

**Comment by MichaelStritt** [comment]
@HenkMutsaerts: You could test if it fails in `xASL_imp_DCM2NII_Subject_ShuffleTheDynamics` already using a breakpoint in line 92:

image

**Comment by HenkMutsaerts** [comment]
> @HenkMutsaerts: You could test if it fails in `xASL_imp_DCM2NII_Subject_ShuffleTheDynamics` already using a breakpoint in line 92:
> 
> image

Thanks, found the issue, `jsonPar.ImageType` was a cell that wasn't unpacked yet.
Fixed this now and will commit.

Though my content is:
    'ORIGINAL'
    'PRIMARY'
    'EPI'
    'NONE '
which wasn't yet in the options.

**Comment by HenkMutsaerts** [comment]
> > @HenkMutsaerts: You could test if it fails in `xASL_imp_DCM2NII_Subject_ShuffleTheDynamics` already using a breakpoint in line 92:
> > image
> 
> Thanks, found the issue, `jsonPar.ImageType` was a cell that wasn't unpacked yet.
> Fixed this now and will commit.
> 
> Though my content is:
> 'ORIGINAL'
> 'PRIMARY'
> 'EPI'
> 'NONE '
> which wasn't yet in the options.
And it still doesn't create `jsonLocal.ASLContext`. It concerns the 2D EPI GE (the first one); does it works for you?

**Comment by HenkMutsaerts** [comment]
I don't understand how `xASL_imp_DCM2NII_Subject_ShuffleTheDynamics` works, this only seems to be able to find `ASLContext` for GE 3D, and also I see no option where it could set "control" or "label". Is this script incomplete or simply not explained well enough ;)

**Comment by jan-petr** [comment]
> I don't understand how `xASL_imp_DCM2NII_Subject_ShuffleTheDynamics` works, this only seems to be able to find `ASLContext` for GE 3D, and also I see no option where it could set "control" or "label". Is this script incomplete or simply not explained well enough ;)

This indeed only works for GE, because for other scanners, we don't know how to do it, it doesn't save this in DICOM, or we didn't have time yet to implement it - it's listed as an issue already :)

**Comment by MichaelStritt** [comment]
Started re-testing the flavors on Windows, just to be 100% sure that everything works correctly. Will post the results tomorrow morning 👍

**Comment by MichaelStritt** [comment]
### Philips_PCASL_3DGRASE_functional

```console
Couldnt load .\TmpConversion\Philips_PCASL_3DGRASE_functional\derivatives\ExploreASL\sub-Patient1\x.mat

ERROR: ASL module terminated for subject 1: sub-Patient1
ans =
    'Error using 'xASL_module_ASL'
     Invalid ASL image
```

### Siemens_PASL_3DGRASE_Prisma_Bsup_AD1

```console
========================================================================
Estimating motion:   
    * - SPM12: spm_realign  --------------------------------------------

        There is not enough overlap in the images to obtain a solution.
         
        Offending image:
        .\TmpConversion\Siemens_PASL_3DGRASE_Prisma_Bsup_AD1\derivatives\ExploreASL\sub-Sub1\ASL_1\ASL4D.nii
         
        Please check that your header information is OK.
        The Check Reg utility will show you the initial
        alignment between the images, which must be
        within about 4cm and about 15 degrees in order
        for SPM to find the optimal solution.

        -----------------------------------------  03:00:56 - 14/05/2021

ERROR: ASL module terminated for subject 1: sub-Sub1
ans =
    'Error using 'spm_realign' line 525
     Insufficient image overlap.
```

### Siemens_PASL_3DGRASE_Prisma_Bsup_AD3

```console
Running new M0 processing method
Masking M0 with structural (pGM+pWM)>0.5 & 70% non-zero sorted intensities

ERROR: ASL module terminated for subject 1: sub-Sub1
ans =
    'Array indices must be positive integers or logical values.
     Error in 'xASL_im_M0ErodeSmoothExtrapolate' line 80
     ThresholdN = SortInt(round(0.7*length(SortInt)));
```

### Siemens_PCASL_multiTI

```console
This assumes that a pre-saturation pulse has been played at the start of the sequence
ERROR: ASL module terminated for subject 1: sub-Sub1
ans =
    'Error using 'zeros'
     Size inputs must be scalar.
     Error in 'xASL_quant_BSupCalculation line 102 
     	SignalPercentageVector = zeros(round(ReadoutTime + max(SliceTime)), 1);
```

### \Siemens_PCASL_volunteer

```console
ERROR: ASL module terminated for subject 1: sub-Sub1
ans =
    'Error using 'xASL_spm_admin line 29 
     pathIn should be a char array
```

**Review by jan-petr — COMMENTED** [review]
> ### Philips_PCASL_3DGRASE_functional
> ```
> Couldnt load .\TmpConversion\Philips_PCASL_3DGRASE_functional\derivatives\ExploreASL\sub-Patient1\x.mat
> 
> ERROR: ASL module terminated for subject 1: sub-Patient1
> ans =
>     'Error using 'xASL_module_ASL'
>      Invalid ASL image
> ```
> 
> ### Siemens_PASL_3DGRASE_Prisma_Bsup_AD1
> ```
> ========================================================================
> Estimating motion:   
>     * - SPM12: spm_realign  --------------------------------------------
> 
>         There is not enough overlap in the images to obtain a solution.
>          
>         Offending image:
>         .\TmpConversion\Siemens_PASL_3DGRASE_Prisma_Bsup_AD1\derivatives\ExploreASL\sub-Sub1\ASL_1\ASL4D.nii
>          
>         Please check that your header information is OK.
>         The Check Reg utility will show you the initial
>         alignment between the images, which must be
>         within about 4cm and about 15 degrees in order
>         for SPM to find the optimal solution.
> 
>         -----------------------------------------  03:00:56 - 14/05/2021
> 
> ERROR: ASL module terminated for subject 1: sub-Sub1
> ans =
>     'Error using 'spm_realign' line 525
>      Insufficient image overlap.
> ```
> 
> ### Siemens_PASL_3DGRASE_Prisma_Bsup_AD3
> ```
> Running new M0 processing method
> Masking M0 with structural (pGM+pWM)>0.5 & 70% non-zero sorted intensities
> 
> ERROR: ASL module terminated for subject 1: sub-Sub1
> ans =
>     'Array indices must be positive integers or logical values.
>      Error in 'xASL_im_M0ErodeSmoothExtrapolate' line 80
>      ThresholdN = SortInt(round(0.7*length(SortInt)));
> ```
> 
> ### Siemens_PCASL_multiTI
> ```
> This assumes that a pre-saturation pulse has been played at the start of the sequence
> ERROR: ASL module terminated for subject 1: sub-Sub1
> ans =
>     'Error using 'zeros'
>      Size inputs must be scalar.
>      Error in 'xASL_quant_BSupCalculation line 102 
>      	SignalPercentageVector = zeros(round(ReadoutTime + max(SliceTime)), 1);
> ```
> 
> ### Siemens_PCASL_volunteer
> ```
> ERROR: ASL module terminated for subject 1: sub-Sub1
> ans =
>     'Error using 'xASL_spm_admin' line 29 
>      pathIn should be a char array
> ```

See above - these are all fine. Issues with data in Flavors - zeroed out images, wrong parameters etc - needs to be fixed later. You can copy paste these in a new issue. But not a real problem in the code.

File: Functions/xASL_bids_determineImageTypeGE.m [hunk 1]
```diff
@@ -23,6 +23,14 @@
     
     % Check ImageType field
     if isfield(jsonPar, 'ImageType')
+        if iscell(jsonPar.ImageType) && length(jsonPar.ImageType)>1
```
[the remark this answers is not in the document]
Comment (jan-petr) [reply Functions/xASL_bids_determineImageTypeGE.m]: I have reverted this one...

**Comment by MichaelStritt** [comment]
> 
> 
> > ### Philips_PCASL_3DGRASE_functional
> > ```
> > Couldnt load .\TmpConversion\Philips_PCASL_3DGRASE_functional\derivatives\ExploreASL\sub-Patient1\x.mat
> > 
> > ERROR: ASL module terminated for subject 1: sub-Patient1
> > ans =
> >     'Error using 'xASL_module_ASL'
> >      Invalid ASL image
> > ```
> > 
> > 
> >   
> >     
> >     
> > 
> >     
> >     
> > 
> >   
> > 
> > 
> > ### Siemens_PASL_3DGRASE_Prisma_Bsup_AD1
> > ```
> > ========================================================================
> > Estimating motion:   
> >     * - SPM12: spm_realign  --------------------------------------------
> > 
> >         There is not enough overlap in the images to obtain a solution.
> >          
> >         Offending image:
> >         .\TmpConversion\Siemens_PASL_3DGRASE_Prisma_Bsup_AD1\derivatives\ExploreASL\sub-Sub1\ASL_1\ASL4D.nii
> >          
> >         Please check that your header information is OK.
> >         The Check Reg utility will show you the initial
> >         alignment between the images, which must be
> >         within about 4cm and about 15 degrees in order
> >         for SPM to find the optimal solution.
> > 
> >         -----------------------------------------  03:00:56 - 14/05/2021
> > 
> > ERROR: ASL module terminated for subject 1: sub-Sub1
> > ans =
> >     'Error using 'spm_realign' line 525
> >      Insufficient image overlap.
> > ```
> > 
> > 
> >   
> >     
> >     
> > 
> >     
> >     
> > 
> >   
> > 
> > 
> > ### Siemens_PASL_3DGRASE_Prisma_Bsup_AD3
> > ```
> > Running new M0 processing method
> > Masking M0 with structural (pGM+pWM)>0.5 & 70% non-zero sorted intensities
> > 
> > ERROR: ASL module terminated for subject 1: sub-Sub1
> > ans =
> >     'Array indices must be positive integers or logical values.
> >      Error in 'xASL_im_M0ErodeSmoothExtrapolate' line 80
> >      ThresholdN = SortInt(round(0.7*length(SortInt)));
> > ```
> > 
> > 
> >   
> >     
> >     
> > 
> >     
> >     
> > 
> >   
> > 
> > 
> > ### Siemens_PCASL_multiTI
> > ```
> > This assumes that a pre-saturation pulse has been played at the start of the sequence
> > ERROR: ASL module terminated for subject 1: sub-Sub1
> > ans =
> >     'Error using 'zeros'
> >      Size inputs must be scalar.
> >      Error in 'xASL_quant_BSupCalculation line 102 
> >      	SignalPercentageVector = zeros(round(ReadoutTime + max(SliceTime)), 1);
> > ```
> > 
> > 
> >   
> >     
> >     
> > 
> >     
> >     
> > 
> >   
> > 
> > 
> > ### Siemens_PCASL_volunteer
> > ```
> > ERROR: ASL module terminated for subject 1: sub-Sub1
> > ans =
> >     'Error using 'xASL_spm_admin' line 29 
> >      pathIn should be a char array
> > ```
> 
> See above - these are all fine. Issues with data in Flavors - zeroed out images, wrong parameters etc - needs to be fixed later. You can copy paste these in a new issue. But not a real problem in the code.

Thanks Jan, then everything should be good to go, right? 👍

**Review by jan-petr — APPROVED** [review]
OK to merge. We need to changes VERSION.md and add issue #570 to CHANGES.md - enough to add it to an existing line about updating headers :)

**Pull request merged**